python/pyspark/sql/types.py

```diff
@@ -187,8 +187,11 @@ def needConversion(self):
 
     def toInternal(self, dt):
         if dt is not None:
-            seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
-                       else time.mktime(dt.timetuple()))
+            seconds = calendar.timegm(dt.utctimetuple())
+            # Avoiding the invalid range of years (100-1899) for mktime in Python < 3
+            if dt.year not in range(100, 1900):
```

Comment:
Thanks for putting this together. I filed the original ticket, but didn't have time to look into how to actually fix it.  This looks like a good fix, but there's a small performance optimization that would help. Changing this to `if 100 >= dt.year > 1900:` gives a significant improvement.

```
# Running 1,000,000 iteration of `toInternal`

Timing of datetime(16,1,1)
------------------------------
Before: 33.7063019276s
After: 1.56738209724s


Timing of datetime(1600,1,1)
--------------------------------
Before: 28.1155149937s
After: 1.55464696884s


Timing of datetime(1900,1,1)
--------------------------------
Before: 33.3729829788s
After: 1.54201412201s
```

With the call to `range`, you have to instantiate the range object for every single serialization (in Python 2 this creates the whole list at once in memory) and then iterate over it in order to do the search. The simple test above avoids all those steps.